---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---

{% include gh_variables.html %}

Attendees will learn how to read, filter, and interrogate variant calls that are in VCF (Variant Call Format) format with Bioconductor packages in R. Analyses may include some or all of the following topics: assessing minor allele frequency, missing data rate, heterozygosity, and/or linkage disequilibrium; working with genome annotations as TxDb objects; discovering what genes are near SNPs of interest; identifying functional consequences of variants.

Before the workshop starts, please follow the [setup](setup.md) instructions.

This workshop was developed by Lindsay Clark at HPCBio, Roy J. Carver
Biotechnology Center, University of Illinois, Urbana-Champaign, using a
website template developed by The Carpentries.

To register to attend this workshop December 2-4, 2020, please see our
[Eventbrite page](https://www.eventbrite.com/e/advanced-r-variant-analysis-fall-2020-registration-129018819763).

For guidelines on how to help improve this lesson,
please see [the contribution guidelines][contributing].

> ## Prerequisites
>
> * **Beginner knowledge of bioinformatics concepts**
>   - Single nucleotide polymorphisms (SNPs)
>   - Genome annotations
>   - Variant and genotype calling
> * **Beginner knowledge of R and Bioconductor**
>   - Installing packages from Bioconductor
>   - Finding help pages
>   - Importing data
>   - Using functions
>   - Indexing and subsetting data
{: .prereq}



{% include links.md %}
