---
source: Rmd
title: "Running statistics on SNP markers"
teaching: 30
exercises: 10
questions:
- "How can I analyze a SNP dataset in R?"
objectives:
- "Calculate allele frequency, missing data rate, heterozygosity, and linkage disequilibrium."
- "Evaluate population structure using principal coordinates analysis"
keypoints:
- "The snpStats package can convert genotypes to numeric format and calculate statistics."
---

```{r, include=FALSE}
source("../bin/chunk-options.R")
knitr_fig_path("04-")
```


```{r, message = FALSE, warning = FALSE}
library(VariantAnnotation)
library(snpStats)
library(ggplot2)
library(scater)
library(magrittr)
```

Picking up where we left off in the last episode:

```{r}
myvcf3 <- VcfFile("data/hmp321_agpv4_chr1_subset_filtered.vcf.bgz")
hdr <- scanVcfHeader(myvcf3)
all_sam <- samples(hdr)
keep_sam <- all_sam[!all_sam %in% c("32", "98F1")]
keep_regions <- GRanges(seqnames = "1",
                        ranges = IRanges(start = c(21.4e6, 22.9e6),
                                         end = c(22.3e6, 23.8e6)))
names(keep_regions) <- c("Region_1", "Region_2")
svp <- ScanVcfParam(info = c("DP", "MAF"), geno = "GT",
                    samples = keep_sam, which = keep_regions,
                    fixed = "ALT")
mydata <- readVcf(myvcf3, param = svp, genome = seqlevels(keep_regions))
```

## Converting from VCF to snpMatrix

The `snpStats` package has a number of functions for analyzing SNP data.  It
uses its own class called `SnpMatrix`. Luckily there is a conversion function
for `CollapsedVCF` objects, which is what we have.

```{r}
mysnpmat <- genotypeToSnpMatrix(mydata)
mysnpmat
```

The `ignore` column indicates markers that were ignored for having multiple
alleles, being insertions or deletions, or being coded in a non-diploid
fashion.  Let's also see how the genotypes are stored.

```{r}
matrix(as.numeric(mysnpmat$genotypes[1:10,1:10]),
       nrow = 10, ncol = 10)
```

In this case, 1 and 3 indicate homozygotes, 2 indicates a heterozygote, and 0
indicates missing data.

> ## Reading from other formats
>
> If you want to import to the snpStats package from other formats such as PED
> or PLINK, see the following vignette:
>
> ```{r, eval = FALSE}
> vignette("data-input-vignette", package = "snpStats")
> ```
>
> You can also use `write.plink` to export to PLINK format.
{: .callout}

## Basic statistics and filtering on SNPs

There's not a lot of point to keeping SNPs that were all set to 0, so let's
eliminate them using the `ignore` column.

```{r}
mat <- mysnpmat$genotypes[,!mysnpmat$map$ignore]
mat
```

The `summary` function give us an overview of the dataset. Here, "rows" are
samples and "cols" are SNPs.

```{r}
summary(mat)
```
 
The "call rate" is the proportion of data that were non-missing.  So, the
samples ranged from 0-16% missing data, and the SNPs ranged from 0-79%
missing data.  100% of calls were certain because we imported from the `GT`
field, but if we had imported from the `GL` field there would have been some
uncertainty.  Each sample has a proportion heterozygosity called, which could
be used to confirm hybrid vs. inbred individuals or identify sample
contamination.  Allele frequencies, per SNP, are expressed both with respect to
the ALT allele (`RAF`) and the minor allele (`MAF`).  We have frequencies of
all three possible genotypes for a diploid.  Lastly we have the z-statistic for
departure from Hardy-Weinberg Equilibrium.

Let's look at per-sample data.

```{r}
sample_stats <- row.summary(mat)

ggplot(sample_stats, aes(x = Call.rate, y = Heterozygosity)) +
  geom_point()
```

We'll discard outliers for heterozygosity.  In your own work, you might
consider filtering by call rate as well.

```{r}
highhet <- isOutlier(sample_stats$Heterozygosity, type = "higher")

ggplot(sample_stats, aes(x = Call.rate, y = Heterozygosity,
                         color = highhet)) +
  geom_point()

mat2 <- mat[!highhet,]
mat2
```

Now we'll look at stats for the markers.

```{r}
marker_stats <- col.summary(mat2)

ggplot(marker_stats, aes(x = MAF)) +
  geom_histogram()
```

The minor allele frequency has a somewhat strange distribution, but that may
have to do with the structure of the various breeding populations used in this
dataset.

Let's look at ratio of observed to expected heterozygosity.

```{r}
marker_stats <- dplyr::mutate(marker_stats, Ho.He = P.AB / (2 * MAF * (1 - MAF)))

ggplot(marker_stats, aes(x = Ho.He)) +
  geom_histogram()
```

We have a few markers that are much more heterozygous than the rest.  Some come out as
`NA` because the minor allele frequency is zero, and we may as well discard those too.

```{r}
highhet2 <- isOutlier(marker_stats$Ho.He, type = "higher")
highhet2[is.na(highhet2)] <- TRUE

mat3 <- mat2[, !highhet2]
mat3
```

> ## Challenge: P.BB vs. RAF
>
> Using our `marker_stats` data frame and `ggplot`, make a graph of `P.BB` vs.
> `RAF`.  What is your interpretation of this plot?  For an extra challenge,
> color the points by `Ho.He` or by `highhet2`.
>
> > ## Solution
> >
> > ```{r}
> > ggplot(marker_stats, aes(x = RAF, y = P.BB)) + geom_point()
> > ggplot(marker_stats, aes(x = RAF, y = P.BB, color = Ho.He)) + geom_point()
> > ggplot(marker_stats, aes(x = RAF, y = P.BB, color = highhet2)) + geom_point()
> > ```
> >
> > Because maize is highly inbred, the frequency of the ALT allele is almost
> > identical the the frequency of homozygotes for the ALT allele.  Markers that
> > differ from that pattern tend to be ones that we identified as being too heterozygous.
> {: .solution}
{: .challenge}

## Linkage Disequilibrium

If we know how much linkage disequilibrium is in our dataset, we know how far
away from a significant SNP to search for candidate genes.  Although not shown
here, it can also be helpful if you want to prune redundant markers.  We'll
calculate it with the `ld` function and visualize the first 500 markers.

```{r}
mydepth <- 100 # how many adjacent markers to look at
myLD <- ld(mat3, depth = mydepth, stats = "R.squared", symmetric = FALSE)
```

```{r}
image(myLD[1:500, 1:500], lwd = 0)
```

There are some loose blocks of LD, but also a lot of adjacent markers that are
not in LD with each other.

We'll get the physical distance between markers based on an approach demonstrated
in `vignette("ld-vignette", package = "snpStats")`.

```{r}
pos <- start(rowRanges(mydata)[colnames(mat3)])
nSNP <- length(pos)
diags <- vector("list", mydepth)
for (i in 1:mydepth) diags[[i]] <- pos[(i+1):nSNP] - pos[1:(nSNP-i)]
physical_distance <- bandSparse(nSNP, k=1:mydepth, diagonals=diags)
```

Now we'll plot LD vs. physical distance.

```{r echo = FALSE}
set.seed(1123) # for report reproducibility
```

```{r plotld}
physical_distance_vals <- physical_distance@x
LD_vals <- myLD@x

random_subset <- sample(which(physical_distance_vals < 2e5), 5000)

ggplot(mapping = aes(x = physical_distance_vals[random_subset],
                     y = LD_vals[random_subset])) +
  labs(x = "Distance in bp", y = "R-squared") +
  geom_point(alpha = 0.1) +
  geom_smooth(formula = y ~ log(x)) +
  geom_vline(xintercept = c(2000, 5000), color = "red", lty = 2)
```

Most but not all LD seems to decay after 2-5 kb.

## Principal components analysis

We should visualize the population structure of the dataset.  This can help
identify groupings that should be accounted for in GWAS or other analysis.  It
can also help to identify groups of samples that are very different from the
rest due to species misidentification or technical issues.

First, we'll use the `xxt` function to fill in missing data and multiply the
genotype matrix by itself transposed.

```{r}
my_xxt <- xxt(mat3)
```

Then, we'll use `eigen` to perform PCA.

```{r}
my_pca <- eigen(my_xxt, symmetric = TRUE)
```

We'll visualize the percentage variation explained by each axis.

```{r}
percent_variation <- round(my_pca$values/sum(my_pca$values) * 100, 2)
plot(percent_variation)
```

The cutoff is arbitrary, but probably at least the first six PCs are worth
investigating.  We'll make a function to plot PCs by number.

```{r, fig.width = 5.5, fig.height = 5}
plotPCs <- function(x, y, eigenvect = my_pca$vectors,
                    pct_var = percent_variation){
  ggplot(mapping = aes(x = eigenvect[,x], y = eigenvect[,y])) +
    geom_point() +
    labs(x = paste0("PC", x, " (", pct_var[x], "%)"),
         y = paste0("PC", y, " (", pct_var[y], "%)"))
}

plotPCs(1, 2)
plotPCs(3, 4)
plotPCs(5, 6)
```

Nothing here is too concerning.  We might want to export the PCA
values and see what individuals get separated out on which axes.

```{r}
pca_tab <- data.frame(Sample = rownames(mat3),
                      my_pca$vectors[,1:6])
colnames(pca_tab)[-1] <- paste0("PC", 1:6)

pca_tab %>% dplyr::filter(PC1 < -0.05) %>%
  dplyr::select(Sample, PC1, PC2)
```

```{r eval = FALSE}
write.csv(pca_tab, file = "maize_pca.csv", row.names = FALSE)
```

> ## Challenge: Find those accessions
>
> PCs 3 and 4 separate out a tight cluster of individuals, in the upper left
> of the plot.  What are the identities of these?
>
> > ## Solution
> > ```{r}
> > pca_tab %>% dplyr::filter(PC3 < -0.17) %>%
> >   dplyr::select(Sample, PC1, PC2, PC3, PC4)
> > ```
> {: .solution}
{: .challenge}

{% include links.md %}
